flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/filesystem/FileSystemTableSource.java

```diff
@@ -314,20 +394,86 @@ public String asSummaryString() {
+    // will influence the final output and physical type used by formats
+    // --------------------------------------------------------------------------------------------
+
+    @Override
+    public void applyProjection(int[][] projectedFields, DataType producedDataType) {
+        this.projectFields = projectedFields;
+        this.producedDataType = producedDataType;
+    }
+
+    @Override
+    public void applyReadableMetadata(List<String> metadataKeys, DataType producedDataType) {
+        if (!metadataKeys.isEmpty()) {
+            this.metadataKeys = metadataKeys;
+            this.producedDataType = producedDataType;
+        }
+    }
+
+    // --------------------------------------------------------------------------------------------
+    // Metadata handling
+    // --------------------------------------------------------------------------------------------
+
+    @Override
+    public Map<String, DataType> listReadableMetadata() {
+        return Arrays.stream(ReadableFileInfo.values())
+                .collect(Collectors.toMap(ReadableFileInfo::getKey, ReadableFileInfo::getDataType));
+    }
+
+    interface FileInfoAccessor extends Serializable {
+        /**
+         * Access the information from the {@link org.apache.flink.core.fs.FileInputSplit}. The
+         * return value type must be an internal type.
+         */
+        Object getValue(FileSourceSplit split);
     }
 
-    private DataType getProjectedDataType() {
-        final DataType physicalDataType = super.getPhysicalDataType();
+    enum ReadableFileInfo implements Serializable {
+        FILEPATH(
+                "filepath",
+                DataTypes.STRING().notNull(),
+                new FileInfoAccessor() {
+                    private static final long serialVersionUID = 1L;
+
+                    @Override
+                    public Object getValue(FileSourceSplit split) {
+                        return StringData.fromString(split.path().getPath());
+                    }
+                });
+
+        final String key;
+        final DataType dataType;
+        final FileInfoAccessor converter;
+
+        ReadableFileInfo(String key, DataType dataType, FileInfoAccessor converter) {
+            this.key = key;
+            this.dataType = dataType;
+            this.converter = converter;
+        }
 
-        // If we haven't projected fields, we just return the original physical data type,
-        // otherwise we need to compute the physical data type depending on the projected fields.
-        if (projectedFields == null) {
-            return physicalDataType;
+        public String getKey() {
+            return key;
+        }
+
+        public DataType getDataType() {
+            return dataType;
+        }
+
+        public FileInfoAccessor getAccessor() {
+            return converter;
+        }
+
+        public static ReadableFileInfo resolve(String key) {
+            switch (key) {
+                case "filepath":
```

Comment:
why having the constant two times? we can simply iterate through ReadableFileInfo values.